pkg/ingestor/parser/spdx/parse_spdx.go

```diff
@@ -0,0 +1,254 @@
+			for _, ext := range pac.PackageExternalReferences {
+				if strings.HasPrefix(ext.RefType, "cpe") {
+					currentPackage.CPEs = append(currentPackage.CPEs, ext.Locator)
+				} else if ext.RefType == spdx_common.TypePackageManagerPURL {
+					currentPackage.Purl = ext.Locator
+				}
+			}
+		}
+		if len(pac.PackageChecksums) > 0 {
+			for _, checksum := range pac.PackageChecksums {
+				currentPackage.Digest = append(currentPackage.Digest, string(checksum.Algorithm)+":"+checksum.Value)
+			}
+		}
+		s.packages[string(pac.PackageSPDXIdentifier)] = append(s.packages[string(pac.PackageSPDXIdentifier)], currentPackage)
+	}
+}
+
+func (s *spdxParser) getFiles() {
+	for _, file := range s.spdxDoc.Files {
+		currentFile := assembler.ArtifactNode{}
+		for _, checksum := range file.Checksums {
+			currentFile.Name = file.FileName
+			currentFile.Digest = string(checksum.Algorithm) + ":" + checksum.Value
+			s.files[string(file.FileSPDXIdentifier)] = append(s.files[string(file.FileSPDXIdentifier)], currentFile)
+		}
+	}
+}
+
+func parseSpdxBlob(p []byte) (*v2_2.Document, error) {
+	reader := bytes.NewReader(p)
+	spdx, err := spdx_json.Load2_2(reader)
+	if err != nil {
+		return nil, err
+	}
+	return spdx, nil
+}
+
+func (s *spdxParser) CreateNodes() []assembler.GuacNode {
+	nodes := []assembler.GuacNode{}
+	for _, packNodes := range s.packages {
+		for _, packNode := range packNodes {
+			nodes = append(nodes, packNode)
+		}
+	}
+	for _, fileNodes := range s.files {
+		for _, fileNode := range fileNodes {
+			nodes = append(nodes, fileNode)
+		}
+	}
+	return nodes
+}
+
+func (s *spdxParser) getPackageElement(elementID string) []assembler.PackageNode {
+	if packNode, ok := s.packages[string(elementID)]; ok {
+		return packNode
+	}
+	return nil
+}
+
+func (s *spdxParser) getFileElement(elementID string) []assembler.ArtifactNode {
+	if fileNode, ok := s.files[string(elementID)]; ok {
+		return fileNode
+	}
+	return nil
+}
+
+func (s *spdxParser) CreateEdges(ctx context.Context, foundIdentities []assembler.IdentityNode) []assembler.GuacEdge {
+	logger := logging.FromContext(ctx)
+	edges := []assembler.GuacEdge{}
+	toplevel := s.getPackageElement("SPDXRef-DOCUMENT")
+	// adding top level package edge manually for all depends on package
+	if toplevel != nil {
+		edges = append(edges, createTopLevelEdges(toplevel[0], s.packages)...)
+	}
+	for _, rel := range s.spdxDoc.Relationships {
+		foundPackNodes := s.getPackageElement("SPDXRef-" + string(rel.RefA.ElementRefID))
+		foundFileNodes := s.getFileElement("SPDXRef-" + string(rel.RefA.ElementRefID))
+		relatedPackNodes := s.getPackageElement("SPDXRef-" + string(rel.RefB.ElementRefID))
+		relatedFileNodes := s.getFileElement("SPDXRef-" + string(rel.RefB.ElementRefID))
+		for _, packNode := range foundPackNodes {
+			createdEdge, err := getEdge(packNode, rel.Relationship, relatedPackNodes, relatedFileNodes)
+			if err != nil {
+				logger.Errorf("error generating spdx edge %v", err)
```

Comment:
wonder if `CreateEdge` should return error... a discussion for a longer term, don't think we need to mind for now,